### PR TITLE
UME-2 with auxiliary tasks

### PR DESCRIPTION
**Pull request by karinazad**
## Description
* Add `UME-2` class for sequence-only encoders 
* Add support for auxiliary training tasks

## Type of Change
- [ ] Bug fix
- [ ] New feature  
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

### CODE REVIEW

File: src/lobster/datasets/s3_datasets/base.py [hunk 1]
```diff
@@ -61,6 +60,7 @@ def __init__(
         seed: int = 0,
         cache_dir: str | None = None,
         transform_fn: Callable | None = None,
```
Comment (taylormjs) [inline src/lobster/datasets/s3_datasets/base.py:62]: Why not have just one `transform_fns` instead of both `transform_fn` and `extra_transform_fns`?

File: src/lobster/hydra_config/experiment/ume-2/base.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      pooling: mean
+      num_layers: 2
+      dropout: 0.1
+  lr: 1e-5
+  scheduler: constant_with_warmup
+  scheduler_kwargs:
+    num_warmup_steps: 10_000
+  weight_decay: 0.01
+  mask_token_id: 8
+  pad_token_id: 6
```
Comment (taylormjs) [inline src/lobster/hydra_config/experiment/ume-2/base.yaml:39]: Why are the masks and pad ids 8 and 6, resp?

File: src/lobster/model/neobert/neobert_module.py [hunk 3]
```diff
@@ -98,3 +100,37 @@ def get_logits(
...
+            return output["last_hidden_state"]
+
+        if not ignore_padding:
+            return output["last_hidden_state"].mean(dim=1)
+
+        mask = attention_mask.to(dtype=output["last_hidden_state"].dtype).unsqueeze(-1)
+
+        masked_embeddings = output["last_hidden_state"] * mask
+
+        sum_embeddings = masked_embeddings.sum(dim=1)
```
Comment (taylormjs) [inline src/lobster/model/neobert/neobert_module.py:133]: In a separate MR, we should probably update this with a pooling function that could be passed to the model. Something like making aggregate (bool) into aggregator (fn)

File: src/lobster/model/ume2/_ume_sequence_encoder.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            raise ValueError(f"Unsupported pooling type: {self.pooling}")
+
+        if self.loss_weight <= 0 or self.loss_weight > 1:
+            raise ValueError(f"Loss weight must be between 0 and 1: {self.loss_weight}")
+
+        if self.task_type not in {"regression"}:
+            raise ValueError(f"Unsupported task type: {self.task_type}")
+
+
+class AuxiliaryRegressionTaskHead(nn.Module):
```
Comment (taylormjs) [inline src/lobster/model/ume2/_ume_sequence_encoder.py:41]: I wonder if this should live elsewhere, like `lobster.finetune` (`lobster.post_train`?). I know we're still pre-training here, but having a dedicated place for pooling, regression heads, etc. that could be used for auxiliary task and post-training might be better organizationally

File: tests/lobster/transforms/functional/test__rdkit_descs.py [hunk 5]
```diff
@@ -15,16 +17,19 @@
...
 @pytest.mark.parametrize(
     ["smiles", "expected"],
     [
         ("foo", None),
         ("CCO", [1.0, 2.0]),
     ],
 )
-def test_smiles_to_rdkit_descs(mock_calc, smiles, expected):
-    mock_calc.return_value = {"desc1": 1.0, "desc2": 2.0}
+def test_smiles_to_rdkit_descs(mock_smiles_to_desc, smiles, expected):
```
Comment (taylormjs) [inline tests/lobster/transforms/functional/test__rdkit_descs.py:28]: Are there tests for ume2 outputting both masked token & auxiliary task preds?

**Review by taylormjs — APPROVED** [review]

File: src/lobster/datasets/s3_datasets/base.py (hunk 1 above)
Comment (karinazad) [reply src/lobster/datasets/s3_datasets/base.py:62]: transform_fn would be applied to the sequence before it goes to tokenization (e.g replace `|` for `.` in protein complexes) and `extra_transform_fns` are applied alongside the tokenized result to give something like

```
{input_ids: ..., attention_mask: ..., extra_output_1: ...., extra_output2:...
```

maybe there is a better name for the parameter though

File: src/lobster/hydra_config/experiment/ume-2/base.yaml (hunk 2 above)
Comment (karinazad) [reply src/lobster/hydra_config/experiment/ume-2/base.yaml:39]: it's a pretty arbitrary order but that's what out tokenizers are using https://github.com/prescient-design/lobster/blob/main/src/lobster/tokenization/_ume_tokenizers.py#L105
